packages/snaps-execution-environments/src/common/endowments/security-utils/object-walker.ts

```diff
@@ -0,0 +1,95 @@
+/**
+ * Object Walker function is testing utility function used to detect when
+ * a certain object reference is present within the provided object.
+ *
+ * Note: Walker will search only for object references, primitive types are omitted.
+ * This function cannot work with very large nested structures because of
+ * the call stack size limit. Further optimizations are required.
+ *
+ * @param subject - Object-like structure to be searched for a reference.
+ * @param targetReference - Target reference.
+ * @returns True if reference to a target is found, false otherwise.
+ */
+export function walkPropertiesAndSearchForReference(
+  subject: unknown,
+  targetReference: unknown,
+) {
+  const seenObjects = new Set();
+
+  /**
+   * Recursively walk properties and search for reference.
+   *
+   * @param currentValue - Object-like structure to be searched for a reference.
+   * @param target - Target reference.
+   * @returns True if reference to a target is found, false otherwise.
+   */
+  function walkAndSearch(currentValue: unknown, target: unknown): boolean {
+    // Check for nulls or undefined and skip further process
+    if (currentValue === undefined || currentValue === null) {
+      return false;
+    }
+    // Check value type and stop process if its a primitive
+    const typeOfValue = typeof currentValue;
+    if (
+      typeOfValue === 'bigint' ||
+      typeOfValue === 'boolean' ||
+      typeOfValue === 'number' ||
+      typeOfValue === 'string' ||
+      typeOfValue === 'symbol'
+    ) {
+      return false;
+    }
+
+    // Circular object detection (handling)
+    // Check if the same object already exists
+    if (seenObjects.has(currentValue)) {
+      return false;
+    }
+    // Add new object to the seen objects set
+    // Only the plain objects should be added (Primitive types are skipped)
+    seenObjects.add(currentValue);
+
+    // TODO: Investigate and find a reason why this is failing
+    //  for some objects, possibly __proto__
+    let objectProperties: [string, any][] = [];
+    try {
+      // Extract object properties
+      objectProperties = Object.entries(currentValue);
```

Comment:
this will only list enumerable properties. 

don't fix, I'll share a different implementation with you

Reply:
Ok, this is resolved now by adding the object-walker from a lavamoat/walker package.